presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java

```diff
@@ -68,27 +62,26 @@
     private final AtomicReference<String> schema = new AtomicReference<>();
     private final AtomicReference<String> timeZoneId = new AtomicReference<>();
     private final AtomicReference<Locale> locale = new AtomicReference<>();
-    private final URI uri;
-    private final HostAndPort address;
+
+    private final PrestoDriverUri prestoDriverUri;
+
     private final String user;
     private final Map<String, String> clientInfo = new ConcurrentHashMap<>();
     private final Map<String, String> sessionProperties = new ConcurrentHashMap<>();
     private final AtomicReference<String> transactionId = new AtomicReference<>();
     private final QueryExecutor queryExecutor;
 
-    PrestoConnection(URI uri, String user, QueryExecutor queryExecutor)
+    PrestoConnection(PrestoDriverUri prestoDriverUri, String user, QueryExecutor queryExecutor)
             throws SQLException
     {
-        this.uri = requireNonNull(uri, "uri is null");
-        this.address = HostAndPort.fromParts(uri.getHost(), uri.getPort());
+        this.prestoDriverUri = prestoDriverUri;
```

Comment:
We should extract out the information needed rather than holding onto the class. That keeps this class easier to reason about and avoids confusing around mutable data like catalog and schema.